File: backend/forge/utils/detect_official.py
"""URL classifier — routes known official-API domains away from browser forging.

Rules (keep in sync with frontend/src/utils/detectOfficial.ts):
  amazon             -> CUSTOM  "amazon"   (covered by the hardcoded amazon core)
  mail.google/gmail  -> OFFICIAL "gmail"   (official Gmail SMTP/API — never forge)
  notion             -> OFFICIAL "notion"  (official Notion API — never forge)
  telegram / t.me    -> OFFICIAL "telegram"
  instagram          -> OFFICIAL "instagram"
  youtube / youtu.be -> OFFICIAL "youtube"
  github             -> OFFICIAL "github"
  slack              -> OFFICIAL "slack"
  anything else      -> CUSTOM  url        (scout + LLM forge, 2 tools per site)
"""

OFFICIAL_KEYWORDS = [
    "mail.google", "gmail", "notion.com", "notion.so",
    "t.me", "telegram.org", "instagram.com", "youtube.com", "youtu.be",
    "github.com", "slack.com"
]


def classify_url(url: str) -> dict:
    url_low = (url or "").lower()
    if "amazon" in url_low:
        return {"type": "CUSTOM", "name": "amazon"}
    if any(k in url_low for k in ["mail.google", "gmail"]):
        return {"type": "OFFICIAL", "name": "gmail"}
    if "notion" in url_low:
        return {"type": "OFFICIAL", "name": "notion"}
    if any(k in url_low for k in ["t.me", "telegram"]):
        return {"type": "OFFICIAL", "name": "telegram"}
    if "instagram" in url_low:
        return {"type": "OFFICIAL", "name": "instagram"}
    if any(k in url_low for k in ["youtube.com", "youtu.be"]):
        return {"type": "OFFICIAL", "name": "youtube"}
    if "github.com" in url_low:
        return {"type": "OFFICIAL", "name": "github"}
    if "slack.com" in url_low:
        return {"type": "OFFICIAL", "name": "slack"}
    return {"type": "CUSTOM", "name": url}

File: frontend/src/utils/detectOfficial.ts
/**
 * URL classifier — mirrors backend/forge/utils/detect_official.py exactly.
 * Official-API domains (Gmail, Notion, Telegram, Instagram, YouTube, GitHub, Slack)
 * must NOT be browser-forged; amazon is custom but covered by the hardcoded amazon core.
 */
export interface UrlVerdict {
  type: "OFFICIAL" | "CUSTOM";
  name: string;
}

export function classifyUrl(url: string): UrlVerdict {
  const urlLow = (url || "").toLowerCase();
  if (urlLow.includes("amazon")) {
    return { type: "CUSTOM", name: "amazon" };
  }
  if (["mail.google", "gmail"].some((k) => urlLow.includes(k))) {
    return { type: "OFFICIAL", name: "gmail" };
  }
  if (urlLow.includes("notion")) {
    return { type: "OFFICIAL", name: "notion" };
  }
  if (["t.me", "telegram"].some((k) => urlLow.includes(k))) {
    return { type: "OFFICIAL", name: "telegram" };
  }
  if (urlLow.includes("instagram")) {
    return { type: "OFFICIAL", name: "instagram" };
  }
  if (["youtube.com", "youtu.be"].some((k) => urlLow.includes(k))) {
    return { type: "OFFICIAL", name: "youtube" };
  }
  if (urlLow.includes("github.com")) {
    return { type: "OFFICIAL", name: "github" };
  }
  if (urlLow.includes("slack.com")) {
    return { type: "OFFICIAL", name: "slack" };
  }
  return { type: "CUSTOM", name: url };
}

/** Sites fully covered by a hardcoded core — no scout / no LLM needed. */
export const CORE_SITE_NAMES = new Set(["amazon"]);
